aggregator/entry.go

```diff
@@ -440,72 +544,192 @@ func (e *Entry) updateMetadatasWithLock(
-				var newElem metricElem
-				switch mu.Type {
-				case metric.CounterType:
-					newElem = e.opts.CounterElemPool().Get()
-				case metric.TimerType:
-					newElem = e.opts.TimerElemPool().Get()
-				case metric.GaugeType:
-					newElem = e.opts.GaugeElemPool().Get()
-				default:
-					return errInvalidMetricType
-				}
-				// NB: The pipeline may not be owned by us and as such we need to make a copy here.
-				key.pipeline = key.pipeline.Clone()
-				if err = newElem.ResetSetData(elemID, storagePolicy, aggTypes, key.pipeline); err != nil {
-					return err
-				}
-				list, err := e.lists.FindOrCreate(storagePolicy.Resolution().Window)
-				if err != nil {
-					return err
-				}
-				newListElem, err := list.PushBack(newElem)
-				if err != nil {
-					return err
-				}
-				newAggregations = append(newAggregations, aggregationValue{key: key, elem: newListElem})
+			listID := standardMetricListID{
+				resolution: storagePolicy.Resolution().Window,
+			}.toMetricListID()
+			var err error
+			newAggregations, err = e.addNewAggregationKey(metric.Type, elemID, key, listID, newAggregations)
+			if err != nil {
+				return err
 			}
 		}
 	}
 
 	// Mark the outdated elements as tombstoned.
-	for _, val := range e.aggregations {
-		if !newAggregations.contains(val.key) {
-			val.elem.Value.(metricElem).MarkAsTombstoned()
-		}
-	}
+	e.removeOldAggregations(newAggregations)
 
 	// Replace the existing aggregations with new aggregations.
 	e.aggregations = newAggregations
 	e.hasDefaultMetadatas = hasDefaultMetadatas
 	e.cutoverNanos = sm.CutoverNanos
-	e.metrics.metadataUpdates.Inc(1)
+	e.metrics.untimed.metadatasUpdates.Inc(1)
 
 	return nil
 }
 
-func (e *Entry) addMetricWithLock(timestamp time.Time, mu unaggregated.MetricUnion) error {
+func (e *Entry) addUntimedWithLock(timestamp time.Time, mu unaggregated.MetricUnion) error {
 	multiErr := xerrors.NewMultiError()
 	for _, val := range e.aggregations {
-		if err := val.elem.Value.(metricElem).AddMetric(timestamp, mu); err != nil {
+		if err := val.elem.Value.(metricElem).AddUnion(timestamp, mu); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+	return multiErr.FinalError()
+}
+
+func (e *Entry) addForwarded(
+	metric aggregated.Metric,
+	metadata metadata.ForwardMetadata,
+) error {
+	timeLock := e.opts.TimeLock()
+	timeLock.RLock()
+
+	// NB(xichen): it is important that we determine the current time
+	// within the time lock. This ensures time ordering by wrapping
+	// actions that need to happen before a given time within a read lock,
+	// so it is guaranteed that actions before when a write lock is acquired
+	// must have all completed. This is used to ensure we never write metrics
+	// for times that have already been flushed.
+	currTime := e.opts.ClockOptions().NowFn()()
+	e.recordLastAccessed(currTime)
+
+	e.RLock()
+	if e.closed {
+		e.RUnlock()
+		timeLock.RUnlock()
+		return errEntryClosed
+	}
+
+	// Reject datapoints that arrive too late.
+	if err := e.checkLateness(
```

Comment:
Could metrics with different `numForwardedTimes` being aggregated together? In that case is it ok for them to have different allowed lateness due to different `numForwardedTimes` in their metadata?

Reply:
sg